enterprise_access/apps/subsidy_access_policy/models.py

```diff
@@ -287,48 +288,101 @@ def will_exceed_spend_limit(self, content_key, content_metadata=None):
     def can_redeem(self, lms_user_id, content_key, skip_customer_user_check=False):
         """
         Check that a given learner can redeem the given content.
+
+        Returns:
+            3-tuple of (bool, str, list of dict):
+                * first element is true if the learner can redeem the content,
+                * second element contains a reason code if the content is not redeemable,
+                * third a list of any transactions represending existing redemptions (any state).
         """
         if not self.active:
-            return (False, REASON_POLICY_NOT_ACTIVE)
+            return (False, REASON_POLICY_NOT_ACTIVE, [])
 
         if not skip_customer_user_check:
             if not self.lms_api_client.enterprise_contains_learner(self.enterprise_customer_uuid, lms_user_id):
-                return (False, REASON_LEARNER_NOT_IN_ENTERPRISE)
+                return (False, REASON_LEARNER_NOT_IN_ENTERPRISE, [])
 
         if not self.catalog_contains_content_key(content_key):
-            return (False, REASON_CONTENT_NOT_IN_CATALOG)
+            return (False, REASON_CONTENT_NOT_IN_CATALOG, [])
 
         subsidy_can_redeem_payload = self.subsidy_client.can_redeem(
             self.subsidy_uuid,
             lms_user_id,
             content_key,
         )
+        existing_transactions = subsidy_can_redeem_payload.get('all_transactions', [])
+
         if not subsidy_can_redeem_payload.get('can_redeem', False):
-            return (False, REASON_NOT_ENOUGH_VALUE_IN_SUBSIDY)
+            return (False, REASON_NOT_ENOUGH_VALUE_IN_SUBSIDY, existing_transactions)
 
         content_metadata = self.get_content_metadata(content_key)
         if not content_metadata:
-            return (False, REASON_CONTENT_NOT_IN_CATALOG)
+            return (False, REASON_CONTENT_NOT_IN_CATALOG, existing_transactions)
 
         if self.will_exceed_spend_limit(content_key, content_metadata=content_metadata):
-            return (False, REASON_POLICY_SPEND_LIMIT_REACHED)
+            return (False, REASON_POLICY_SPEND_LIMIT_REACHED, existing_transactions)
+
+        return (True, None, existing_transactions)
+
+    def _redemptions_for_idempotency_key(self, all_transactions):
+        """
+        Select the historical redemptions (transactions) that may contribute to the idempotency key.
+
+        Currently, only failed or reversed transactions qualify.  That means for every new failed/reversed transaction
+        in the history, a new transaction with a unique idempotency_key can be created even for the same
+        subsidy/policy/learner/content.  In all other cases, the same idempotency_key as the last redeem attempt will be
```

Comment:
I would just add something here like: "thus allowing us to re-attempt a previously failed or reversed transaction for that same (subsidy, policy, learner, content) combination."